File: aws/dump-student-dnsrr.pl
#!/usr/bin/perl
#
# Dump DNS resource records for all student instances

require 5.6.0;
use strict;
use warnings;
use Data::Dumper;
use JSON -support_by_pp;

my %instance_id       = ();
my %instance_hostname = ();
my %instance_type     = ();

my $profile = "default";
my $region  = "eu-west-1";
my $awscmd = sprintf("aws --profile %s --region %s", $profile, $region);

my $json = `$awscmd ec2 describe-instances`;
my $data = decode_json($json);

foreach my $r (@{ $data->{Reservations} }) {
    foreach my $i (@{ $r->{Instances} }) {

        my $addr = $i->{PublicIpAddress};

        my $name = undef;
        my $id   = undef;
        my $type = undef;

        foreach my $t (@{ $i->{Tags} }) {
            $id   = $t->{Value} if ($t->{Key} eq "ID");
            $type = $t->{Value} if ($t->{Key} eq "Type");
            $name = $t->{Value} if ($t->{Key} eq "Name");
        }

	next unless $type;
	next unless $addr;

        if ($type eq "teacher") {
            printf("teacher IN A %s\n", $addr);
            next;
        }

        if ($type eq "resolver") {
            printf("resolver%-12d IN A %s\n", $id, $addr);
            next;
        }

        if ($type eq "signer") {
            printf("ns%-18d IN A %s\n",       $id, $addr);
            printf("group%-15d IN NS ns%d\n", $id, $id);
            next;
        }
    }
}

File: aws/dump-student-dnsrr.py
"""Extract instance DNS RRs"""

import boto3

ec2 = boto3.client(service_name="ec2")

filters = [{"Name": "tag:Type", "Values": ["signer", "resolver"]}]
response = ec2.describe_instances(Filters=filters)

for r in response.get("Reservations"):
    for i in r.get("Instances"):
        public_ip_address = i.get("PublicIpAddress")

        instance_type = None
        instance_id = None

        for t in i.get("Tags"):
            if t.get("Key") == "Type":
                instance_type = t.get("Value")
            if t.get("Key") == "ID":
                instance_id = t.get("Value")

        if instance_id is not None:
            if instance_type == "teacher":
                print(f"teacher{instance_id} IN A {public_ip_address}")
            elif instance_type == "resolver":
                print(f"resolver{instance_id} IN A {public_ip_address}")
            elif instance_type == "signer":
                print(f"ns{instance_id} IN A {public_ip_address}")
                print(f"group{instance_id} IN NS ns{instance_id}")
